Show support links in Help Center instead of modal

klimeryk opened the pull request:
As suggested in pbvpgB-3o7-p2

If a user clicks on a support doc link in Calypso, a new modal (or tab) is opened with it. This takes the user away from the context they’re need that information the most. Thanks to Vertex's work on Help Center, we can instead show those pages the Help Center. This will keep the user in the Calypso context, while giving them a better UX. The floating window of Help Center can be moved around and stays as they navigate Calypso and, for example, follow the steps on the support doc.

I've used the existing `InlineSupportLink` component that is used across Calypso (though, not consistently). It has been introduced in 2018 in https://github.com/Automattic/wp-calypso/pull/26347, and fits perfectly our needs.

This change will make the support doc modal and related code obsolete - I'll remove it in a separate PR, once we verify this approach does not have any negative side-effects or use cases we did not consider.

While working on this, I noticed a regression - Help Center disappears when the view is changed (eg. you go from Home to Upgrades). I've created a separate PR to address that: https://github.com/Automattic/wp-calypso/pull/79742.

## Proposed Changes

* Use Help Center instead of a modal to display support docs.

## Testing Instructions

- Find a support link (usually it's the "Learn more" ones) and click on it - it should open the Help Center and its contents should be loaded.
- Verify that the `Back` button in Help Center works as expected - takes you to the root of its navigation.

klimeryk commented on packages/wpcom-proxy-request/src/index.js:
```diff
@@ -15,7 +15,7 @@ const proxyOrigin = 'https://public-api.wordpress.com';
  * "Origin" of the current HTML page.
  */
-const origin = window.location.protocol + '//' + window.location.host;
+const origin = globalThis?.location?.protocol + '//' + globalThis?.location?.host;
```
@jsnajdr, I'd love your (or someone else's from Team Calypso) expert opinion on this. If I don't make this change, it will trigger an "Undefined  error while trying to render SSR, it seems:

```
/Users/klimeryk/work/workspace/calypso/build/webpack:/wp-calypso/packages/wpcom-proxy-request/src/index.js:18
const origin = window?.location?.protocol + '//' + window?.location?.host;
               ^
ReferenceError: window is not defined
    at Object../client/components/inline-support-link/index.jsx (/Users/klimeryk/work/workspace/calypso/build/webpack:/wp-calypso/packages/wpcom-proxy-request/src/index.js:18:16)
    at __webpack_require__ (/Users/klimeryk/work/workspace/calypso/build/webpack:/wp-calypso/webpack/bootstrap:19:1)
    at Object../client/my-sites/plugins/index.node.js (/Users/klimeryk/work/workspace/calypso/build/server.js:45587:27)
    at __webpack_require__ (/Users/klimeryk/work/workspace/calypso/build/webpack:/wp-calypso/webpack/bootstrap:19:1)
    at Object.load (/Users/klimeryk/work/workspace/calypso/build/webpack:/wp-calypso/client/sections.js:146:17)
    at forEach (/Users/klimeryk/work/workspace/calypso/build/webpack:/wp-calypso/client/server/pages/index.js:908:13)
    at Array.forEach (<anonymous>)
    at pages (/Users/klimeryk/work/workspace/calypso/build/webpack:/wp-calypso/client/server/pages/index.js:899:4)
    at boot (/Users/klimeryk/work/workspace/calypso/build/webpack:/wp-calypso/client/server/boot/index.js:96:11)
    at /Users/klimeryk/work/workspace/calypso/build/webpack:/wp-calypso/client/server/index.js:14:13
```

Looks like this line is what ultimately leads to loading of wpcom-proxy-request in this context (since `@automattic/data-stores` uses internally `wpcom-proxy-request`, IIRC):

https://github.com/Automattic/wp-calypso/pull/79741/files#diff-8b10c9eab3990cb2c7310ada64772c069aafd016f76fc7692328b5f148cd6a98R2

I found this related Calypso PR https://github.com/Automattic/wp-calypso/pull/65576, which led to a suggestion to use `globalThis` from you (p4TIVU-a9o-p2#comment-11513).

alshakero replied:
This solution is fine for sure, but it's a bit of a bandage.

`origin` is only used twice, both don't need to be in module initialization. I'd inline `origin` to where it is used, and then it can stay `window`. 

Maybe move `debug( 'using "origin": %o', origin )` inside `install` and use `window.location.protocol + '//' + window.location.host` right away.

klimeryk replied:
Thanks for the nudge - I thought about it initially, but wanted to keep the changes to minimum and avoid changes to behaviour. I was concerned that there's some edge cases where the value at the time of module initialization will differ from the one inline. But it makes more sense and is clearer, so let's try it 💪 

I've extracted these changes to a separate PR: https://github.com/Automattic/wp-calypso/pull/79789. That way it's easier to test, approve and revert if needed.

alshakero commented on packages/wpcom-proxy-request/src/index.js:
```diff
@@ -85,7 +85,7 @@ const requests = {};
  * See: http://goo.gl/xxYf6D
  */
-const supportsProgress = !! window.ProgressEvent && !! window.FormData;
+const supportsProgress = !! globalThis.ProgressEvent && !! globalThis.FormData;
```
I'd inline this next to the line that needs it

```js
params.supports_progress = supportsProgress; 
```

klimeryk committed:
Show support links in Help Center instead of modal

As suggested in pbvpgB-3o7-p2

If a user clicks on a support doc link in Calypso, a new modal (or tab) is opened with it. This takes the user away from the context they’re need that information the most. Thanks to Vertex's work on Help Center, we can instead show those pages the Help Center. This will keep the user in the Calypso context, while giving them a better UX. The floating window of Help Center can be moved around and stays as they navigate Calypso and, for example, follow the steps on the support doc.

I've used the existing InlineSupportLink component that is used across Calypso (though, not consistently). It has been introduced in 2018 in #26347, and fits perfectly our needs.

klimeryk committed:
Fix e2e tests

AtrumGeost reviewed (commented):
Thanks for working on this. I can already say this will offer a better experience, having the doc in Help Center allows to navigate (follow the steps) through the interface 🎉  I tested it and it works as expected for the main purpose/case 👍

I also tested opening another link when a doc is already opened, _though not sure if this will be a "usual case"_. The link won't open while help center is opened:

https://github.com/Automattic/wp-calypso/assets/3696121/daf0bacd-bb23-47e6-815e-384ec63c63ff

---

> I've used the existing InlineSupportLink component that is used across Calypso (though, not consistently).

Based on the above comment I think maybe the following is not within the scope of this PR but sharing just in case: not all support links will open inside the Help Center, for instance for the four links in this screenshot, one will open in a new tab:

![CleanShot 2023-07-26 at 11 19 05](https://github.com/Automattic/wp-calypso/assets/3696121/b912f99b-4f55-4686-8c4c-1a035bbc6216)

The same with the **Recommended resources** links at the bottom of the **My Home** screen:

![CleanShot 2023-07-26 at 11 21 19](https://github.com/Automattic/wp-calypso/assets/3696121/50fd09ec-7d55-4c01-a8e5-c6da14443000)

klimeryk committed:
Ensure navigation works in edge and not-so-edge cases

Props to Jorge!

klimeryk commented:
Thank you for thoroughly testing, @AtrumGeost, appreciate it!

> I also tested opening another link when a doc is already opened, though not sure if this will be a "usual case". The link won't open while help center is opened:

That's a great point and a case I completely missed. It's pretty severe so I went and fixed it, as without it, it could be a confusing behaviour (Help Center open, support links are not working).

See https://github.com/Automattic/wp-calypso/pull/79741/commits/fdb551cc11c600a57b3e01c14d1de26b2a1a5ec1

> Based on the above comment I think maybe the following is not within the scope of this PR but sharing just in case: not all support links will open inside the Help Center, for instance for the four links in this screenshot, one will open in a new tab:

Yes, those are on purpose left as-is. It's out of scope of this PR, but I'll look into converting them in the future (and welcome help if you're up for some Calypso work).

klimeryk commented on packages/data-stores/src/help-center/actions.ts:
```diff
@@ -130,6 +129,16 @@ export const setShowMessagingChat = function* () {
+	const params = new URLSearchParams( {
+		link,
+		postId: String( postId ),
+		cacheBuster: String( Date.now() ),
```
This unused field is there, as the name implies, to "bust" "cache" - without it, if the user clicks on a support link, Help Center opens, etc. But if they navigate (back, to a different doc, etc.) within Help Center, clicking that support link in the main interface will not do anything (because the `initialRoute` value will be the same - no state change, no `useEffect` trigger). This will ensure it's always triggered without the need to revamp how we store the value or using more complicated approaches. But open to suggestions if there's a better option I missed 🙇‍♂️

dpasque commented:
Thanks for updating the E2E tests! 😄 From a KitKat perspective, looks good 👍 -- I'll defer the final approval to the review iterations in progress!

AtrumGeost reviewed (approved):
> That's a great point and a case I completely missed. It's pretty severe so I went and fixed it, as without it, it could be a confusing behaviour (Help Center open, support links are not working).

Cool! I tested it again and now it works 🎉 

> Yes, those are on purpose left as-is. It's out of scope of this PR, but I'll look into converting them in the future (and welcome help if you're up for some Calypso work).

Sure! It would be beneficial to me more familiarity with Calypso 👍

simison commented:
👏  Excellent to see this!

> > Based on the above comment I think maybe the following is not within the scope of this PR but sharing just in case: not all support links will open inside the Help Center, for instance for the four links in this screenshot, one will open in a new tab:
> 
> Yes, those are on purpose left as-is. It's out of scope of this PR, but I'll look into converting them in the future (and welcome help if you're up for some Calypso work).

Another good follow-up could be to split support articles a bit smaller. For [example having wp-admin](https://wordpress.com/support/settings/writing-settings/#writing-settings-in-the-wp-admin-dashboard) and calypso versions, for multiple separate sections within each article is a bit much. With the modal approach, less is better. E.g. wp-admin and calypso sections could be in separate articles, each feature/section have its own article, etc.

cc @donalirl
